bindings/node/test/providers/credentialsProvider.test.js

```diff
@@ -0,0 +1,527 @@
+        const url = httpSpy.args[0][0];
+        expect(url).to.be.instanceof(URL);
+        expect(url.searchParams.get('resource'), 'https://vault.azure.net');
+      });
+
+      it('sends the request to `http://169.254.169.254/metadata/identity/oauth2/token`', () => {
+        const url = httpSpy.args[0][0];
+        expect(url).to.be.instanceof(URL);
+        expect(url.toString()).to.include('http://169.254.169.254/metadata/identity/oauth2/token');
+      });
+
+      it('sets the Metadata header to true', () => {
+        const options = httpSpy.args[0][1];
+        expect(options).to.have.property('headers').to.have.property('Metadata', true);
+      });
+
+      it('sets the Content-Type header to application/json', () => {
+        const options = httpSpy.args[0][1];
+        expect(options)
+          .to.have.property('headers')
+          .to.have.property('Content-Type', 'application/json');
+      });
+
+      context('prose test specific requirements', () => {
+        /**
+         * the driver prose tests require the ability to set custom URL endpoints
+         * for the IMDS call and set custom headers
+         */
+        const url = new URL('http://customentpoint.com');
+
+        beforeEach(async () => {
+          sinon.restore();
+          httpSpy = sinon.stub(utils, 'get');
+          httpSpy.resolves(mockResponse);
+          await fetchAzureKMSToken({
+            url,
+            headers: {
+              customHeader1: 'value1',
+              customHeader2: 'value2'
+            }
+          });
+        });
+
+        it('allows a custom URL to be specified', () => {
+          const url = httpSpy.args[0][0];
+          expect(url).to.be.instanceof(URL);
+          expect(url.toString()).to.include('http://customentpoint.com');
+        });
+
+        it('deep copies the provided url', () => {
+          const spiedUrl = httpSpy.args[0][0];
+          expect(spiedUrl).to.be.instanceof(URL);
+          expect(spiedUrl).to.not.equal(url);
+        });
+
+        it('allows custom headers to be specified', () => {
+          const options = httpSpy.args[0][1];
+          expect(options).to.have.property('headers').to.have.property('customHeader1', 'value1');
+          expect(options).to.have.property('headers').to.have.property('customHeader2', 'value2');
+        });
+      });
+    });
+
+    context('error handling', () => {
+      afterEach(() => sinon.restore());
+      context('when the request times out', () => {
+        before(() => {
+          sinon.stub(utils, 'get').rejects(new MongoCryptKMSRequestNetworkTimeoutError());
+        });
+
+        it('throws a MongoCryptKMSRequestError', async () => {
+          const error = await loadCredentials({ azure: {} }).catch(e => e);
+          expect(error).to.be.instanceOf(MongoCryptAzureKMSRequestError);
+        });
+      });
+
+      context('when the request returns a non-200 error', () => {
+        context('when the request has no body', () => {
```

Comment:
might also add a case for when a request has a non-json body

Reply:
done!